ui/src/app/store/utils/slice.ts

```diff
@@ -0,0 +1,345 @@
+      },
+    },
+    updateStart: (state: GenericState<I, E>) => {
+      state.saved = false;
+      state.saving = true;
+    },
+    updateError: (state: GenericState<I, E>, action: PayloadAction<E>) => {
+      state.errors = action.payload;
+      state.saving = false;
+    },
+    updateSuccess: (state: GenericState<I, E>) => {
+      state.errors = null;
+      state.saved = true;
+      state.saving = false;
+    },
+    updateNotify: (state: GenericState<I, E>, action: PayloadAction<I>) => {
+      for (const i in state.items) {
+        if (state.items[i].id === action.payload.id) {
+          state.items[i] = action.payload;
+        }
+      }
+    },
+    delete: {
+      // Slices that use a different key e.g. system_id should overwrite this
+      // action and reducer.
+      prepare: (id: I["id"]) => ({
+        meta: {
+          model: name,
+          method: "delete",
+        },
+        payload: {
+          params: {
+            id,
+          },
+        },
+      }),
+      reducer: () => {
+        // No state changes need to be handled for this action.
+      },
+    },
+    deleteStart: (state: GenericState<I, E>) => {
+      state.saved = false;
+      state.saving = true;
+    },
+    deleteError: (state: GenericState<I, E>, action: PayloadAction<E>) => {
+      state.errors = action.payload;
+      state.saving = false;
+    },
+    deleteSuccess: (state: GenericState<I, E>) => {
+      state.errors = null;
+      state.saved = true;
+      state.saving = false;
+    },
+    deleteNotify: (
+      state: GenericState<I, E>,
+      action: PayloadAction<I["id"]>
+    ) => {
+      const index = state.items.findIndex(
+        (item: I) => item.id === action.payload
+      );
+      state.items.splice(index, 1);
+    },
+    cleanup: (state: GenericState<I, E>) => {
+      state.errors = null;
+      state.saved = false;
+      state.saving = false;
+    },
+  };
+  return createSlice({
+    initialState: {
+      errors: null,
+      items: [],
+      loaded: false,
+      loading: false,
+      saved: false,
+      saving: false,
+      ...initialState,
+    },
+    name,
+    reducers: {
+      ...baseReducers,
+      ...reducers,
+    },
+    // Due to types being lost we need to cast to the slice type.
+    // See: https://github.com/reduxjs/redux-toolkit/issues/580
+  }) as Slice<GenericState<I, E>, R & typeof baseReducers, typeof name>;
+};
+
+/**
+ * A utility to generate reducers and actions to append to a slice.
+ * @template S - A model that includes status e.g. Machine.
+ * @template I - A model that is used as an an array of items on the provided
+ *               state e.g. DHCPSnippet
+ * @template K - A model key e.g. "id"
+ * @param {string} name - The name of the model that matches the name in MAAS.
+ * @param {string} indexKey - The key used to index a model e.g. "id"
+ *                            or "system_id".
+ * @param {array} handlers - A collection of status handlers.
```

Comment:
I wasn't quite sure about this one as it's quite a complex object. It could be `Object[]` or I could use a full `@typedef` (https://jsdoc.app/tags-typedef.html). But not sure if that's worth it as there is a full type for the object with its own descriptions...

Any thoughts about which way to go?